### PR TITLE
Picking System

### PR DESCRIPTION
**Pull request by roughleythomas**
It picks sort of

### CODE REVIEW

File: Engine/src/Engine/SceneBuilder/Components.h
Comment (Snow106) [inline Engine/src/Engine/SceneBuilder/Components.h:179]: would be nice to have a comment here with "Axis Aligned Bounding Box" so other people looking at this struct can get a better understanding (or at least an easier time googling). Just a suggestion

File: Engine/src/Engine/Rendering/UIRenderer.cpp
Comment (Snow106) [inline Engine/src/Engine/Rendering/UIRenderer.cpp]: these are part of another PR right? maybe omit these from here for now, or wait until the other PR is approved since it has changes pending

File: Engine/src/Engine/Rendering/Renderer.cpp
Comment (Snow106) [inline Engine/src/Engine/Rendering/Renderer.cpp]: brackets

File: Engine/src/Engine/Rendering/Renderer.cpp
Comment (Snow106) [inline Engine/src/Engine/Rendering/Renderer.cpp]: brackets

File: Engine/src/Engine/Physics/OctTree.h
Comment (Snow106) [inline Engine/src/Engine/Physics/OctTree.h]: Some spacing would be nice between each of these

File: Engine/src/Engine/Physics/OctNode.h
Comment (Snow106) [inline Engine/src/Engine/Physics/OctNode.h:11]: just a suggestion, I think it's a bit more readable if you make one comment for all forward declarations

File: Engine/src/Engine/Physics/OctTree.h
Comment (Snow106) [inline Engine/src/Engine/Physics/OctTree.h]: can be forward declared

File: Engine/src/Engine/Physics/OctNode.cpp
Comment (Snow106) [inline Engine/src/Engine/Physics/OctNode.cpp]: is this meant to be commented?

**Review by Snow106 — CHANGES_REQUESTED** [review]

File: Engine/src/Engine/EngineUI/GamePanel.cpp
Comment (KevinAlexV) [inline Engine/src/Engine/EngineUI/GamePanel.cpp]: Brace should be on a newline, iirc

File: Engine/src/Engine/Physics/AABB.h
Comment (KevinAlexV) [inline Engine/src/Engine/Physics/AABB.h:13]: Why is LTF the only enum with 0 as the instantiated value? This is likely just my lower understanding of C++.

File: Engine/src/Engine/Physics/OctTree.h
Comment (KevinAlexV) [inline Engine/src/Engine/Physics/OctTree.h]: I think a little more clarification here would be good. The world is divided into 8 portions that each handle a set of physics checks? Or is this to handle something else?

File: Engine/src/Engine/Physics/Ray.h
Comment (KevinAlexV) [inline Engine/src/Engine/Physics/Ray.h:6]: What's ray used for? Is it just for the mouse click location?

File: Engine/src/Engine/Rendering/UIRenderer.cpp
Comment (KevinAlexV) [reply Engine/src/Engine/Rendering/UIRenderer.cpp]: These ones in particular, I don't think were included in the other PR. More an @Hevano question though

File: Engine/src/Engine/Rendering/UIRenderer.cpp
Comment (KevinAlexV) [inline Engine/src/Engine/Rendering/UIRenderer.cpp]: Bracket on newline - but this can also be moved into the header, if desired. Might be cleaner, but just a suggestion!

File: Engine/src/Engine/SceneBuilder/Components.h
Comment (KevinAlexV) [reply Engine/src/Engine/SceneBuilder/Components.h:179]: I'd also think that some brief notes on how to use the component would be incredibly useful. I need to do that for the components I created as well 😅

File: Engine/src/Engine/SceneBuilder/InputSystem.cpp
Comment (KevinAlexV) [inline Engine/src/Engine/SceneBuilder/InputSystem.cpp]: Braces on newline

**Review by KevinAlexV — COMMENTED** [review]
Insane work, thanks a bunch Thomas!

Made some comments, if you don't mind taking a look

File: Engine/src/Engine/SceneBuilder/InputSystem.cpp
Comment (roughleythomas) [reply Engine/src/Engine/SceneBuilder/InputSystem.cpp]: Done

File: Engine/src/Engine/Rendering/UIRenderer.cpp
Comment (roughleythomas) [reply Engine/src/Engine/Rendering/UIRenderer.cpp]: Moved to header

File: Engine/src/Engine/Physics/Ray.h
Comment (roughleythomas) [reply Engine/src/Engine/Physics/Ray.h:6]: The one class I forgot to comment. If its still vague please ping.

File: Engine/src/Engine/Rendering/Renderer.cpp
Comment (roughleythomas) [reply Engine/src/Engine/Rendering/Renderer.cpp]: Fixed

File: Engine/src/Engine/Physics/OctTree.h
Comment (roughleythomas) [reply Engine/src/Engine/Physics/OctTree.h]: Tried to make it more clear what an oct-tree is. If its still unclear, might just remove the explanation

File: Engine/src/Engine/Physics/AABB.h
Comment (roughleythomas) [reply Engine/src/Engine/Physics/AABB.h:13]: You can assign integer/hex values to enums and reference them as such. In this instance LTF is the first corner/cube in the order of traversal so it is index 0, RBB is the final corner/cube and as such is index 7. This is used in a lot of the AABB referencing code.

File: Engine/src/Engine/Rendering/UIRenderer.cpp
Comment (Snow106) [reply Engine/src/Engine/Rendering/UIRenderer.cpp]: Ill approve the PR thenbut I do recommend to rename this function to getPosition instead of getPos, just for consistency

**Review by Snow106 — DISMISSED** [review]

File: Engine/src/Engine/Physics/AABB.h
Comment (KevinAlexV) [reply Engine/src/Engine/Physics/AABB.h:13]: Ok awesome, thanks for the explanation!

**Review by Snow106 — APPROVED** [review]

**Review by KevinAlexV — APPROVED** [review]
Looks good!

**Pull request merged**
